### ISSUE TITLE
✨ Feature – Add support for catch-all routes

### ISSUE THREAD

**Issue by ryannhg** [elm-land/elm-land#24]
## Problem

Sometimes, the structure of a URL isn't fixed. Dynamic routes allow users to handle URLs like `/users/1` and `/users/2` easily with a single page. But for an application like GitHub, the length of the URL isn't known.

For example, viewing a file should be possible on a single page:

```
https://github.com/elm-land/elm-land/blob/main/README.md
https://github.com/elm-land/elm-land/blob/main/examples/01-hello-world/elm-land.json
https://github.com/elm-land/elm-land/blob/main/examples/01-hello-world/src/Pages/Home_.elm
```

## Solution

Allow users to specify "catch-all" routes that handle URLs with an unknown depth. Here's how you could support the GitHub file preview shown in the problem section above:

```
src/Pages/Owner_/Repo_/Blob/Branch_/ALL_.elm
```

That would provide the following URL parameters for use in the application:

```elm
module Pages.Owner_.Repo_.Blob.Branch_.ALL_ exposing (Model, Msg, page)

type alias Params =
    { owner : String
    , repo : String
    , branch : String
    , first_ : String
    , rest_ : List String
    }
    
page : Shared.Model -> Route Params -> Page Model Msg
page shared route =
     ...
```

When a user visits a URL like `/elm-land/elm-land/blob/main/examples/01-hello-world/elm-land.json`

This will be the value available in `route.params`:

```elm
route.params ==
    { owner = "elm-land"
    , repo = "elm-land"
    , branch = "main"
    , first_ = "examples"
    , rest_ = [ "01-hello-world", "elm-land.json" ]
    }
```

__Note:__ The dynamic catch-all parameters are split between `route.params.first_` and `route.params.rest_`, so that you are never in the impossible state of having an empty list. This page would NOT handle requests to `/elm-land/elm-land/blob/main`, because there are no URL paths after the branch name.

## More examples

Page | URL | Elm file | Example URLs
:-- | :-- | :-- | :--
Homepage | `/` | `src/Pages/Home_.elm` | `/`
Sign in | `/sign-in` | `src/Pages/SignIn.elm` | `/sign-in`
Settings | `/settings/account` | `src/Pages/Settings/Account.elm` | `/settings/account`
Profile | `/profile/:username` | `src/Pages/Profile/Username_.elm` | `/profile/ryan` <br/> `/profile/duncan` <br/> ...
Blog Landing | `/blog` | `src/Pages/Blog.elm` | `/blog`
Blog Post Detail | `/blog/*` | `src/Pages/Blog/ALL_.elm` | `/blog/hello-world` <br/> `/blog/elm-land/installation` <br/> `/blog/cooking/lasagna`  <br/> ...

```elm
module Pages.Blog.ALL_exposing (page)

import Html exposing (Html)
import View exposing (View)


page : { first_ : String, rest_: List String } -> View msg
page params =
    { title = "Pages.Blog.ALL_"
    , body = [ Html.text "/blog/*" ]
    }
```

**Comment by ryannhg** [comment]
This morning, an Elm Land Discord user pointed out that the name `STAR_.elm` isn't very clear to them. 

They made some great suggestions for alternative names that would be more straightforward. ( thank you `voxal` )

- ALL
- ANY
- WILDCARD
- CATCH_ALL

To help make a decision, I did some quick research on how other popular JS frameworks name files with this behavior. This is a helpful way to make sure that newcomers aren't surprised!

Here's what I learned:
1. [NuxtJS](https://nuxtjs.org/docs/features/file-system-routing#unknown-dynamic-nested-routes)
    - Calls them "Unknown dynamic nested routes"
    - `pages/blog/_.vue`
2. [NextJS](https://nextjs.org/docs/routing/dynamic-routes#catch-all-routes)
    - Calls them "Catch-all routes"
    - `pages/blog/[...slug].jsx`
3. [Astro](https://docs.astro.build/en/core-concepts/routing/#rest-parameters)
    - Calls them "Rest parameters"
    - `pages/blog/[...slug].astro`
4. [SvelteKit](https://kit.svelte.dev/docs/advanced-routing#rest-parameters)
    - Calls them "Rest parameters"
    - `routes/blog/[...slug].svelte`
5. [Gatsby](https://www.gatsbyjs.com/docs/reference/routing/file-system-route-api/#splat-routes)
    - Calls them "Splat routes"
    - `pages/blog/[...slug].js`

The challenge with making an Elm framework is that the name of a module is rather restricted:
- Must start with a capital letter (A–Z)
- Can only contain letters, numbers, and underscores

To avoid confusion with the existing static routes and dynamic routes, it feels like using an all caps name, followed by an underscore (to denote it's a special route) is the best path forward.

Here's where I'm leaning at the moment:

- Call them "Catch-all routes"
- `Pages/Blog/ALL_.elm`

( I'm leaning towards `ALL_.elm` because it fits with the name "Catch-all", but isn't too verbose! )

**Comment by ryannhg** [comment]
As another note, the official `elm/url` parser isn't designed for parsing paths of unknown length.

Instead, for our use case we can just use a normal `case` expression in our generated code.

See this Ellie example for an interactive example of the snippet below:
https://ellie-app.com/jSXSnXKf5vba1

```elm
fromUrl : Url -> Path
fromUrl url =
    let
        urlPathSegments : List String
        urlPathSegments =
            url.path
                |> String.split "/"
                |> List.filter (String.isEmpty >> not)
    in
    case urlPathSegments of
        [] ->
            Home_

        owner_ :: [] ->
            Owner_
                { owner = owner_
                }

        owner_ :: repo_ :: [] ->
            Owner__Repo_
                { owner = owner_
                , repo = repo_
                }

        owner_ :: repo_ :: "blob" :: branch_ :: first_ :: rest_ ->
            Owner__Repo__Blob_Branch__ALL_
                { owner = owner_
                , repo = repo_
                , branch = branch_
                , first_ = first_
                , rest_ = rest_
                }

        _ ->
            NotFound_
```

**Comment by ryannhg** [comment]
We also need an algorithm to make sure the route paths are correctly sorted from specific -> general. 

Here's an Ellie with an algorithm that should make sure routes are parsed as expected:
https://ellie-app.com/jSYCWDsTy8ra1

**Comment by ryannhg** [comment]
This has been added in #54 , and will be released with `v0.18.0` 🚀
